### PR TITLE
Samba4: wip recipe

### PR DESCRIPTION
**Pull request by extrowerk**
fails to build:

> [2443/3780] Compiling lib/tdr/testsuite.c
> [2444/3780] Compiling lib/tevent/testsuite.c
> ../lib/tevent/testsuite.c: In function 'test_event_fd2_sock_handler':
> ../lib/tevent/testsuite.c:507:28: error: 'PIPE_BUF' undeclared (first use in this function)
>   if (oth_sock->num_read >= PIPE_BUF) {
>                             ^
> ../lib/tevent/testsuite.c:507:28: note: each undeclared identifier is reported only once for each function it appears in
> Waf: Leaving directory `/sources/samba-4.5.5/bin'

### CODE REVIEW

**Comment by miqlas** [comment]
MD5Init, MD5Update and MD5Final undefined.

**Comment by miqlas** [comment]
Compiles just fine, but i had to move the python modules manually to get it packaged.

Samba comes with plenty private libs, i had to list them in the provides/requires section, i don't know, why. Help wanted!

After installing the package i get the following eror if i tryto run any binary from the package:
`runtime_loader: /boot/system/lib/libsmbconf.so.0: Version dependency "libsamba-cluster-support-samba4.so" not found.`

No idea, what's wrong here.

**Comment by korli** [comment]
See https://github.com/haiku/haiku/blob/master/src/system/runtime_loader/elf_versioning.cpp#L180
Indeed it doesn't look good.

**Comment by miqlas** [comment]
Ok, then it is not because the runtime_loader cannot locate the .so, but there is something wrong with the library let me say so "format or structure".
Could it be some kind of strange compiling option like PIC or stack-protector or is it something real deep?
I've not experienced any strange behaviour during the compiling, but i take plenty things granted, so take my oppinions with a big bunch of salt.

**Comment by miqlas** [comment]
@korli, here you go: https://dl.dropboxusercontent.com/u/844783/samba_problematic_libs.zip

File: net-fs/samba/patches/samba-4.5.5.patchset [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+--- a/lib/tevent/testsuite.c
++++ b/lib/tevent/testsuite.c
+@@ -36,6 +36,13 @@
+ #include <assert.h>
+ #endif
+ 
++#ifdef __HAIKU__
++#include <posix/limits.h>
++#ifndef PIPE_BUF
++#define PIPE_BUF _POSIX_PIPE_BUF
```
Comment (korli) [inline net-fs/samba/patches/samba-4.5.5.patchset]: It seems Haiku defined this value here https://github.com/haiku/haiku/blob/master/headers/private/system/vfs_defs.h#L29
Probably needs to be patched in Haiku

Comment (korli) [inline net-fs/samba/patches/samba-4.5.5.patchset]: Added in hrev51109

**Comment by korli** [comment]
The runtime_loader failure is tracked here: https://dev.haiku-os.org/ticket/13436

About the recipe layout, I have some local changes to avoid private libs appearing directly in $libDir, so I will amend this PR once the runtime_loader bug is fixed.

**Comment by miqlas** [comment]
@korli: thank! Does it works for you?

**Comment by korli** [comment]
It fails to find anything windows based on the local network, so I can't test.

**Comment by miqlas** [comment]
It doesn't really packages on my laptop:
"Warning: POLICY ERROR: can't find used library "libsmbd-shim-samba4.so"
Warning: POLICY ERROR: can't find used library "libtalloc.so.2"
Warning: POLICY ERROR: can't find used library "libsamba-debug-samba4.so"
Warning: POLICY ERROR: can't find used library "libtevent.so.0"
Warning: POLICY ERROR: can't find used library "libsamba-security-samba4.so""

Latest haiku AMD64.

**Comment by korli** [comment]
Yeah some private libs are missing a rpath, so it fail the policy check. But the package provides this lib in a private path.

**Comment by miqlas** [comment]
Hope it helps:

# smbd -i -F -S outputs:
....
messaging_dgm_init: socket failed: Address family not supported by protocol family
messaging_dgm_ref failed: Address family not supported by protocol family

**Comment by korli** [comment]
@miqlas does smbd even work with samba3? 

@orangejua should we keep samba3 as a dependency for fusesmb? If yes, the recipe needs an update to depend on libsmbclient == 3.6.25.

**Comment by korli** [comment]
Ticket 13436 marked as fixed. this should work from hrev51110.

**Comment by miqlas** [comment]
I got fusesmb compiled with samba4, but as samba4 not in working state yet, cannot test it.

**Comment by miqlas** [comment]
Samba3 smbd not tested, but i can check it later.

**Comment by miqlas** [comment]
Tested samba3 smbd, and however i was not able to share anything (sigtermed as i tried to connect),  i havent met an errlor like with smbd4.
The sigterm could happen because my config Files.

**Comment by orangejua** [comment]
@korli Yeah, since I don't think Samba4 can be made to work with gcc2, I'd like to keep fusesmb on Samba3 for the time being.

**Comment by waddlesplash** [comment]
> Tested samba3 smbd, and however i was not able to share anything (sigtermed as i tried to connect), i havent met an errlor like with smbd4.

So does that mean that smbd4 is working and this is ready to merge?

**Comment by miqlas** [comment]
Nope

**Comment by orangejua** [comment]
Just a note: I have a working Samba4 now (although there still seem to be a few issues.) Working on a new SMB-add-on with it, will push then. Just so there's no duplicated work on this :-)

**Comment by miqlas** [comment]
Great news @orangejua!

**Comment by extrowerk** [comment]
@juafromspace : any news here?

**Comment by scottmc** [comment]
ping?

**Comment by extrowerk** [comment]
So here is the startlog: https://gist.github.com/extrowerk/00eb825f47753714d71fa7c45a455ec0

I think the problem is the famous "Address family not supported by protocol family".
No idea how to fix however.

**Comment by Begasus** [comment]
With latest changes pushed to @extrowerk 's PR I get output from above mentioned command:
```
~> smbd -i -F
INFO: Current debug levels:
  all: 5
  tdb: 5
  printdrivers: 5
  lanman: 5
  smb: 5
  rpc_parse: 5
  rpc_srv: 5
  rpc_cli: 5
  passdb: 5
  sam: 5
  auth: 5
  winbind: 5
  vfs: 5
  idmap: 5
  quota: 5
  acls: 5
  locking: 5
  msdfs: 5
  dmapi: 5
  registry: 5
  scavenger: 5
  dns: 5
  ldb: 5
  tevent: 5
...
Paths:
   SBINDIR: /packages/samba4-4.20.2-1/.self/bin
   BINDIR: /packages/samba4-4.20.2-1/.self/bin
   CONFIGFILE: /packages/samba4-4.20.2-1/.self/settings/samba/smb.conf
   LOGFILEBASE: /packages/samba4-4.20.2-1/.self/var/log
   LMHOSTSFILE: /packages/samba4-4.20.2-1/.self/settings/samba/lmhosts
   LIBDIR: /packages/samba4-4.20.2-1/.self/lib
   DATADIR: /packages/samba4-4.20.2-1/.self/data
   SAMBA_DATADIR: /packages/samba4-4.20.2-1/.self/data/samba
   MODULESDIR: /packages/samba4-4.20.2-1/.self/lib/samba
   SHLIBEXT: so
   LOCKDIR: /var/samba/lock
   STATEDIR: /var
   CACHEDIR: /var/cache/samba
   PIDDIR: /var/samba
   SMB_PASSWD_FILE: /packages/samba4-4.20.2-1/.self/settings/samba/smbpasswd
   PRIVATE_DIR: /packages/samba4-4.20.2-1/.self/settings/samba
   BINDDNS_DIR: /packages/samba4-4.20.2-1/.self/var/lib/samba/bind-dns

System Headers:
   HAVE_SYS_CDEFS_H
   HAVE_SYS_FILE_H
   HAVE_SYS_IOCTL_H
   HAVE_SYS_IPC_H
   HAVE_SYS_MMAN_H
   HAVE_SYS_PARAM_H
   HAVE_SYS_RESOURCE_H
   HAVE_SYS_SELECT_H
   HAVE_SYS_SOCKET_H
...
A LOT MORE OUTPUT
...
   vfs_syncops_init
   vfs_time_audit_init
   vfs_unityed_media_init
   vfs_virusfilter_init
   vfs_widelinks_init
   vfs_worm_init
   vfs_xattr_tdb_init

Cluster support features:
   NONE

Type sizes:
   sizeof(char):         1
   sizeof(int):          4
   sizeof(long):         8
   sizeof(long long):    8
   sizeof(uint8_t):      1
   sizeof(uint16_t):     2
   sizeof(uint32_t):     4
   sizeof(short):        2
   sizeof(void*):        8
   sizeof(size_t):       8
   sizeof(off_t):        8
   sizeof(ino_t):        8
   sizeof(dev_t):        4

Builtin modules:
   vfs_default vfs_not_implemented auth_builtin auth_sam auth_winbind nss_info_template idmap_tdb idmap_passdb idmap_nss idmap_ldap
get_current_groups: user is in 2 groups: 101, 0
messaging_dgm_init: socket failed: Address family not supported by protocol family
messaging_dgm_ref failed: Address family not supported by protocol family
```

**Comment by waddlesplash** [comment]
Those `messaging_dgm` failures shouldn't happen on a nightly build.

**Comment by Begasus** [comment]
No users created.
First run:
```
   vfs_widelinks_init
   vfs_worm_init
   vfs_xattr_tdb_init

Cluster support features:
   NONE

Type sizes:
   sizeof(char):         1
   sizeof(int):          4
   sizeof(long):         8
   sizeof(long long):    8
   sizeof(uint8_t):      1
   sizeof(uint16_t):     2
   sizeof(uint32_t):     4
   sizeof(short):        2
   sizeof(void*):        8
   sizeof(size_t):       8
   sizeof(off_t):        8
   sizeof(ino_t):        8
   sizeof(dev_t):        4

Builtin modules:
   vfs_default vfs_not_implemented auth_builtin auth_sam auth_winbind nss_info_template idmap_tdb idmap_passdb idmap_nss idmap_ldap
get_current_groups: user is in 2 groups: 101, 0
directory_create_or_exist: mkdir failed on directory /var/samba/lock: No such file or directory
```

Second run (created directory /var/samba):
```
dbwrap_lock_order_unlock: release lock order 1 for /var/registry.tdb
tdb(/var/registry.tdb): tdb_transaction_start: nesting 1
dbwrap_lock_order_lock: check lock order 1 for /var/registry.tdb
dbwrap_lock_order_unlock: release lock order 1 for /var/registry.tdb
dbwrap_lock_order_lock: check lock order 1 for /var/share_info.tdb
dbwrap_lock_order_unlock: release lock order 1 for /var/share_info.tdb
dbwrap_lock_order_lock: check lock order 1 for /var/share_info.tdb
dbwrap_lock_order_unlock: release lock order 1 for /var/share_info.tdb
Finding user nobody
Trying _Get_Pwnam(), username as lowercase is nobody
Trying _Get_Pwnam(), username as uppercase is NOBODY
Checking combinations of 0 uppercase letters in nobody
Get_Pwnam_internals didn't find user [nobody]!
SamInfo3_for_guest: Unable to locate guest account [nobody]!
get_guest_info3 failed with NT_STATUS_NO_SUCH_USER
ERROR: failed to setup guest info.
```

**Pull request closed without merging**